powershell: fix the build

yrashk opened the pull request:
For some reason, building powershell derivation started failing with
these errors:

```
  liblttng-ust.so.0 -> not found!
  ...
  libpam.so.0 -> not found!
```

This patch adds these dependencies

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

matthewbauer commented:
I think you will need to add it to the library path for this to fix your problem.

yrashk commented:
@matthewbauer I can do that, however currently .so's referring to these libraries resolve well:

```
~/P/n/nixpkgs (powershell-fix|…) $ ldd /nix/store/wb1nbgl7kav2wdsy6hbb20ngznwpbzma-powershell-6.0.2/share/powershell/*.so | grep lttng
	liblttng-ust.so.0 => /nix/store/08ziz6wz4gznjmn5j3n7g9jk5ydk7r3j-lttng-ust-2.10.1/lib/liblttng-ust.so.0 (0x00007fef06584000)
	liblttng-ust-tracepoint.so.0 => /nix/store/08ziz6wz4gznjmn5j3n7g9jk5ydk7r3j-lttng-ust-2.10.1/lib/liblttng-ust-tracepoint.so.0 (0x00007fef054c7000)
~/P/n/nixpkgs (powershell-fix|…) $ ldd /nix/store/wb1nbgl7kav2wdsy6hbb20ngznwpbzma-powershell-6.0.2/share/powershell/*.so | grep pam
	libpam.so.0 => /nix/store/bx17bcz9z9gh7bjh22vi3qk3s8206izw-linux-pam-1.3.0/lib/libpam.so.0 (0x00007fed8c8a6000)
	libpam.so.0 => /nix/store/bx17bcz9z9gh7bjh22vi3qk3s8206izw-linux-pam-1.3.0/lib/libpam.so.0 (0x00007f10ec332000)
```
Do you think I need to cover that in `makeLibraryPath` anyway?

yrashk committed:
powershell: fix the build

For some reason, building powershell derivation started failing with
these errors:

```
  liblttng-ust.so.0 -> not found!
  ...
  libpam.so.0 -> not found!
```

This patch adds these dependencies

yrashk commented:
Amended the patch to include all libraries in the wrapper.

Mic92 commented on pkgs/shells/powershell/default.nix:
```diff
@@ -22,7 +23,7 @@ stdenv.mkDerivation rec {
 
   buildInputs = [ autoPatchelfHook makeWrapper ];
-  propagatedBuildInputs = [ libunwind libuuid icu curl cacert less openssl ];
+  propagatedBuildInputs = [ cacert less ] ++ libraries;
```
Why are all those packages in propagatedBuildInputs?

yrashk replied:
Perhaps I don't fully understand that part yet. The intention here was to submit that those are runtime dependencies, and not explosively build-time ones. If there's another, correct way to do this, please let me know.

Mic92 replied:
Nix figures out runtime dependencies by inspecting what files are referenced by a derivation.
Sometimes it cannot do this if for example an program do PATH search to run a different program.
In this case we use for example makeWrapper to extend the programs PATH.
`propagatedBuildInputs` will only be pulled if it this package is used as an build input for a different package. 
This is also the case for nix-shell by accident, which is why people sometimes think `propagatedBuildInputs` affect the runtime dependencies. Python is an exception to this rule since it uses `propagatedBuildInputs` in its own wrapper to extend the PYTHONPATH to contain all required python libraries used in an python script.

Mic92 replied:
less is already covered by the PAGER variable. Where is cacert used in powershell?

yrashk replied:
As for `PAGER`, this is because there's a bug handling PAGER with arguments in the current release and there's a comment in this in the original nix file for PowerShell:

```nix
# TODO: remove PAGER after upgrading to v6.1.0-preview.1 or later as it has been addressed in
# https://github.com/PowerShell/PowerShell/pull/6144
```

Without this setting, current version will fail commands like `Help`:

```
PS /home/yrashk/Projects/nix/nixpkgs> help                                                                                                                                                                          
Get-Command : The term 'less -R' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

Once 6.1 is released, this can be removed.

yrashk replied:
I've simplified build inputs to make this a much cleaner derivation as per your guidance.

Mic92 commented on pkgs/shells/powershell/default.nix:
```diff
@@ -31,7 +32,7 @@ stdenv.mkDerivation rec {
     cp -r * $out/share/powershell
     rm $out/share/powershell/DELETE_ME_TO_DISABLE_CONSOLEHOST_TELEMETRY
-    makeWrapper $out/share/powershell/pwsh $out/bin/pwsh --prefix ${platformLdLibraryPath} : "${stdenv.lib.makeLibraryPath [ libunwind libuuid icu openssl curl ]}" \
+    makeWrapper $out/share/powershell/pwsh $out/bin/pwsh --prefix ${platformLdLibraryPath} : "${stdenv.lib.makeLibraryPath libraries}" \
```
Are the packages opened with dlopen? platformLdLibraryPath is probably only required on macOS, where there is no patchelf.

yrashk replied:
Yes, they are not explicitly linked to this libraries, that's why this approach was used in the first place.

This patch doesn't change that, just adds more libraries that /for some reason/ weren't an issue before.

Mic92 committed:
powershell:  autoPatchelfHook/makeWrapper belongs to nativeBuildInputs
